### PR TITLE
Add max size, null check, and update for toxext API change, add size negotiation

### PR DESCRIPTION
**Pull request by anthonybilinski**

### CODE REVIEW

File: tox_extension_messages.h
Comment (sudden6) [inline tox_extension_messages.h]: Even though this is a generous limit, I'm quite certain if we hardcode this limit here, we'll sooner or later run into problems again.

I'd prefer to make this configurable for the extension and use your value as default.

Comment (anthonybilinski) [reply tox_extension_messages.h]: Yeah, agreed that it would be better for it to be configurable. I'll add that.

Comment (anthonybilinski) [reply tox_extension_messages.h]: Hmm.. configuring the receive side buffer could cause some bad user experience, e.g. if the two friends have different configured buffers, friend A with size 4 and friend B with size 5, B might keep sending 5-length messages to A who silently drops them.

It seems like each friend should probably know the receive buffer size of their friend and avoid sending over that, which would require some additional negotiation. That size would then need to be exposed up to the user of the client, per-friend, so that they can appropriately limit message size per friend.

Comment (anthonybilinski) [reply tox_extension_messages.h]: To summarize chats with @sudden6 and @sphaerophoria, @sudden6 favours having a constant size and if the size needs to be increased in the future, a extension with a new UUID would be released with the larger constant size. A client could then support both extensions for backwards compatibility, assuming the second extension changed all symbol names as well to avoid conflicts, with the main argument being simplicity of extension implementation and of client implementation (which wouldn't need a different max size for each friend), since increasing the size may either never be needed or not be needed for many years.

@sphaerophoria favours adding the negotiation of receive buffer size per friend, since it would remove the need for a full copy of the extension to change the size, allows more customization per client, and a cleaner upgrade path, and is more conceptually "correct".

Given that this is a tox related project, idealistic solutions seem fitting. I think adding the negotiation wouldn't be too tough, on the toxext callback the extension can then do its own negotiation, and once that's complete call the client callback with the max size, which doesn't seem too complicated.

Thinking about the upgrade path though, if any extension protocol change other than size is needed, in general extensions couldn't do anything clever to support backwards compatibility since they would have a new UUID, and toxext would completely separate them. For e.g. messages, that would mean max size would drop back down to ~1KB and maybe the client could nag the user to nag their friend to update their extension, or the client could link both the old and new extensions then do backwards compatibility itself.. I wonder if there's much to be gained from allowing extensions to version themselves. I guess any extension can always add its own protocol version and use that rather than making a new UUID on protocol changes, so toxext wouldn't need to fundamentally support it. @sphaerophoria what do you think about adding a protocol version to `messenges`? Should that just be handled by either other extensions that wrap it, or by full UUID changes instead?

**Comment by anthonybilinski** [comment]
I've updated with a configurable max size that can be set by the user only before negotiation, and which defaults to the previous value. tox_extension_messages now does its own negotiation on toxext negotiation completion. Max size of all negotiated friends are stored and checked on send with errors returned to caller if exceeded, and on the receiving side if we still get a message that exceeds our negotiating max receiving size, we drop the message.

I added a test covering as much of the few new API as I could think of, but currently have a bug, I'm not sure if in the test or in the impl. Will look into it more tomorrow.

**Comment by anthonybilinski** [comment]
They were legit bugs, yay tests :tada: 

Updated now, ended up making pretty significant changes.. I broke the incoming message parsing into each discrete option since I started handling MESSAGE_PART separately so that we don't have to copy in data until out incoming_message is full before rejecting for that reason, and the parse function was getting pretty unruly IMO.

Ready for review again.

File: tox_extension_messages.c
Comment (sphaerophoria) [inline tox_extension_messages.c]: I think we should remove this and just add our message size to the registration function. It looks like we don't support changing this after negotiation has started and it seems likely that a client will accidentally race this.

File: tox_extension_messages.c
Comment (sphaerophoria) [inline tox_extension_messages.c:354]: I can't remember the design here, but I think it's worth a double check if clients are expecting negotiation to timeout. If toxext gives up at some point and sends a false compatible flag this could introduce a new case when the user goes offline between negotiation and receiving the size negotiation packet.

File: tox_extension_messages.c
Comment (sphaerophoria) [inline tox_extension_messages.c]: I know you're just following the pattern, but can we change these ifs to else ifs just to make it clearer to the reader that it can only be one of the types at a time? At first I couldn't figure out why they didn't return early (it's cause of the iterator update at the bottom).

File: tox_extension_messages.c
Comment (sphaerophoria) [inline tox_extension_messages.c]: Seems kinda messed up to assert if a packet has some unexpected data in it.

**Review by sphaerophoria — CHANGES_REQUESTED** [review]
@anthonybilinski told me to hold off on reviewing this until history is cleaned up a bit, but I already had some comments so I might as well publish them so they can be addressed during the cleanup.

File: tox_extension_messages.c
Comment (anthonybilinski) [reply tox_extension_messages.c]: Agreed, makes sense. Was just doing it that way to max the default val easier to use, but the added code and possible race definitely isn't worth it.

File: tox_extension_messages.c
Comment (anthonybilinski) [reply tox_extension_messages.c]: Commit ordering cleaned up and the whole if ladder is now a switch-case.

File: tox_extension_messages.c
Comment (anthonybilinski) [reply tox_extension_messages.c]: Definitely, bug. Removed.

File: tox_extension_messages.c
Comment (anthonybilinski) [reply tox_extension_messages.c:354]: It looks like toxext only calls the negotiation callback with false  when revoking or when it gets an explicit negotiation response saying the friend doesn't support the extension, so I don't think we're at risk of toxext timing out then us stalling the callback to clients forever.

File: tox_extension_messages.c
Comment (sphaerophoria) [reply tox_extension_messages.c]: This if ladder isn't

File: tox_extension_messages.c
Comment (sphaerophoria) [inline tox_extension_messages.c]: Something about this feels off

1. Why are we copying data in if we know we're dropping it?
2. This is duplicated with the section above. Seems like it's happening unconditionally. I think if friend_data->sending_invalid we should just unconditionally clear the message and return.

**Review by sphaerophoria — CHANGES_REQUESTED** [review]

**Comment by sphaerophoria** [comment]
Looks mostly good, I noticed that we had a lot of arrays of friend_id + something which seemed like a waste. I've pushed a commit to consolidate them all into one "FriendData" struct which should simplify things a lot

```
 2 files changed, 86 insertions(+), 174 deletions(-)
```

Left a couple comments in the review but all in all this looks really good. Thanks for the cleanup and addition.

I ran the new changes under `ctest -T Memcheck` (which initially revealed an uninitialized variable bug) and have no errors/warnings, so I think the new allocation flow is reasonably correct.

*Note*: I didn't really look at the additional tests or the cleanup code too closely. I should probably loop back for a second look before merging

File: test/max_message_test.c
Comment (sphaerophoria) [inline test/max_message_test.c:107]: This is pretty weird because this should be an empty packet list. Indicates that we should probably add a toxext_packetlist_free that allows the caller to abort the send. That being said I don't think there's anything wrong here so I'm not going to bother requesting any changes here.

We _could_ check that the last_received_message_size doesn't change but I'm not going to block this on that since this has been in review for so long and I don't think it's wrong.

**Review by sphaerophoria — APPROVED** [review]

**Comment by sphaerophoria** [comment]
Alright I've done a second pass, updated a couple spots and approved. @anthonybilinski let me know if you want to approve my additional changes or if we should just wrap this up and update the qtox pr

**Pull request merged**
